### PR TITLE
[Evaluation] Dropped 'StateT' from the CEK machine

### PR DESCRIPTION
**Pull request by effectfully**
Don't look here yet.

### CODE REVIEW

**Comment by effectfully** [comment]
I compared using `STRef` vs `StateT` for the same algorithm (i.e. no unpacking of state into two separate variables) and here are the results:

```
crowdfunding/1           3.438 ms	3.324 ms    -3.3%
crowdfunding/2           3.490 ms	3.374 ms    -3.3%
crowdfunding/3           3.492 ms	3.377 ms    -3.3%
crowdfunding/4           2.093 ms	2.027 ms    -3.2%
crowdfunding/5           2.095 ms	2.030 ms    -3.1%
future/1                 2.120 ms	2.063 ms    -2.7%
future/2                 4.056 ms	3.946 ms    -2.7%
future/3                 4.050 ms	3.950 ms    -2.5%
future/4                 5.413 ms	5.272 ms    -2.6%
future/5                 6.381 ms	6.215 ms    -2.6%
future/6                 5.450 ms	5.311 ms    -2.6%
future/7                 6.406 ms	6.216 ms    -3.0%
multisigSM/1             4.001 ms	3.878 ms    -3.1%
multisigSM/2             4.018 ms	3.898 ms    -3.0%
multisigSM/3             4.059 ms	3.930 ms    -3.2%
multisigSM/4             4.096 ms	4.004 ms    -2.2%
multisigSM/5             4.765 ms	4.624 ms    -3.0%
multisigSM/6             3.998 ms	3.875 ms    -3.1%
multisigSM/7             4.021 ms	3.896 ms    -3.1%
multisigSM/8             4.064 ms	3.924 ms    -3.4%
multisigSM/9             4.099 ms	4.004 ms    -2.3%
multisigSM/10            4.773 ms	4.625 ms    -3.1%
vesting/1                3.507 ms	3.382 ms    -3.6%
vesting/2                3.097 ms	3.003 ms    -3.0%
vesting/3                3.506 ms	3.378 ms    -3.7%
marlowe/trustfund/1      11.01 ms	10.70 ms    -2.8%
marlowe/trustfund/2      8.750 ms	8.505 ms    -2.8%
marlowe/zerocoupon/1     11.19 ms	10.95 ms    -2.1%
marlowe/zerocoupon/2     8.050 ms	7.853 ms    -2.4%
```

There was no variance in either benchmark results. So even just by replacing `StateT` with a single `STRef` we gain some performance benefit.

Next I'll try using two `STRef`s.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs [hunk 1]
```diff
@@ -158,19 +158,19 @@ type CekValEnv uni fun = UniqueMap TermUnique (CekValue uni fun)
 -- defers to the function stored in the environment). This makes the budgeting machinery extensible
 -- and allows us to separate budgeting logic from evaluation logic and avoid branching on the union
 -- of all possible budgeting state types during evaluation.
-newtype CekBudgetSpender cost uni fun = CekBudgetSpender
+newtype CekBudgetSpender cost uni fun s = CekBudgetSpender
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: I know you said not to look, but... why not:

```
unCekBudgetSpender :: forall term s . ToExMemory term => STRef ExBudget -> ExBudgetCategory fun -> ExBudget -> CekCarryingM cost term uni fun s()
```
i.e. your spender needs to get passed the STRef when you use it, but that seems fine?

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: The point being, you avoid all the `forall s . ExBudgetMode .. s` stuff, and it's a bit more obvious that in the main implementation we're just passing a `STRef` around and feeding it into this.

Also it'd be easy to get out the final state: just look into the ref!

Comment (effectfully) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: > Also it'd be easy to get out the final state: just look into the ref!

It's designed to be extended to work with two refs (restricting) as well as a single one (counting, tallying).

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: Ah right, I see the issue. I guess I was thinking about what Ed was suggesting with GHC unboxing `STRef`s into mutable variables, which might be inhibited by this kind of abstraction, versus just passing `STRef`s as function arguments.

Comment (effectfully) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: > Ah right, I see the issue. I guess I was thinking about what Ed was suggesting with GHC unboxing STRefs into mutable variables, which might be inhibited by this kind of abstraction, versus just passing STRefs as function arguments.

I don't really know how a mutable variable looks like... Do you happen to have a reference? In my solution all the `STRef`s are within a single definition, so it's kinda like encapsulation which might help optimizations. Or might not indeed.

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]:
```
data STRef s a = STRef (MutVar# s a)
```
`MutVar` is the actual primitive type. So I guess if you want to get all the way down to `MutVar s Int` or whatever, which is presumably about as performant as it's possible to get in Haskell, you need to persuade Haskell to unbox the `STRef`, which means probably worker-wrapper, which means having them hanging around as function args, I think.

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: But I'm guessing based on some off-hand remarks, we can also worry about it later.

Comment (effectfully) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:161]: Ah, I see. Well, I'd expect that optimization to fire more reliably when the `STRef` is not visible to the outside at all and indeed here's what I see when I compile a version with two `STRef`s:

```
-- RHS size: {terms: 110, types: 236, coercions: 85, joins: 0/1}
UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.$wrestricting
  :: forall s (uni :: * -> *) fun.
     ExCPU
     -> ExMemory
     -> GHC.Prim.State# s
     -> (# GHC.Prim.State# s, ExBudgetMode RestrictingSt uni fun s #)
UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.$wrestricting
  = \ (@ s)
      (@ (uni :: * -> *))
      (@ fun)
      (ww
         :: ExCPU
         Unf=OtherCon [])
      (ww1
         :: ExMemory
         Unf=OtherCon [])
      (w :: GHC.Prim.State# s) ->
      case GHC.Prim.newMutVar# ww w of { (# ipv, ipv1 #) ->
      case GHC.Prim.newMutVar# ww1 ipv of { (# ipv2, ipv3 #) ->
      (# ipv2,
         UntypedPlutusCore.Evaluation.Machine.Cek.Internal.ExBudgetMode
           ((\ (@ term)
               _
               _
               (eta1 :: ExBudget)
               _
               (eta3 :: GHC.Prim.State# s) ->
               case eta1 of { ExBudget cpuToSpend memToSpend ->
               case GHC.Prim.readMutVar# ipv1 eta3 of { (# ipv4, ipv5 #) ->
               case GHC.Prim.readMutVar# ipv3 ipv4 of { (# ipv6, ipv7 #) ->
               let {
                 nt :: Integer
                 nt
                   = integer-gmp-1.0.3.0:GHC.Integer.Type.minusInteger
                       (ipv5 `cast` <Co:1>) (cpuToSpend `cast` <Co:1>) } in
               case GHC.Prim.writeMutVar# ipv1 (nt `cast` <Co:2>) ipv6 of s2#
               { __DEFAULT ->
               case integer-gmp-1.0.3.0:GHC.Integer.Type.minusInteger
                      (ipv7 `cast` <Co:1>) (memToSpend `cast` <Co:1>)
               of nt1
               { __DEFAULT ->
               case GHC.Prim.writeMutVar# ipv3 (nt1 `cast` <Co:2>) s2# of s2#1
               { __DEFAULT ->
               case integer-gmp-1.0.3.0:GHC.Integer.Type.ltInteger#
                      nt
                      UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.restricting2
               of {
                 __DEFAULT ->
                   case integer-gmp-1.0.3.0:GHC.Integer.Type.ltInteger#
                          nt1
                          UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.restricting2
                   of {
                     __DEFAULT ->
                       (# s2#1,
                          UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode.counting2 #);
                     1# ->
                       (# s2#1,
                          Data.Either.Left
                            (PlutusCore.Evaluation.Machine.Exception.ErrorWithCause
                               (PlutusCore.Evaluation.Machine.Exception.UserEvaluationError
                                  (UntypedPlutusCore.Evaluation.Machine.Cek.Internal.CekOutOfExError
                                     ((PlutusCore.Evaluation.Machine.ExBudget.ExBudget
                                         (nt `cast` <Co:2>) (nt1 `cast` <Co:2>))
                                      `cast` <Co:2>)))
                               GHC.Maybe.Nothing) #)
                   };
                 1# ->
                   (# s2#1,
                      Data.Either.Left
                        (PlutusCore.Evaluation.Machine.Exception.ErrorWithCause
                           (PlutusCore.Evaluation.Machine.Exception.UserEvaluationError
                              (UntypedPlutusCore.Evaluation.Machine.Cek.Internal.CekOutOfExError
                                 ((PlutusCore.Evaluation.Machine.ExBudget.ExBudget
                                     (nt `cast` <Co:2>) (nt1 `cast` <Co:2>))
                                  `cast` <Co:2>)))
                           GHC.Maybe.Nothing) #)
               }
               }
               }
               }
               }
               }
               })
            `cast` <Co:56>)
           ((\ (s1 :: GHC.Prim.State# s) ->
               case GHC.Prim.readMutVar# ipv1 s1 of { (# ipv4, ipv5 #) ->
               case GHC.Prim.readMutVar# ipv3 ipv4 of { (# ipv6, ipv7 #) ->
               (# ipv6,
                  (PlutusCore.Evaluation.Machine.ExBudget.$WExBudget ipv5 ipv7)
                  `cast` <Co:5> #)
               }
               })
            `cast` <Co:4>) #)
      }
      }
```

`STRef`s do not even get created, there's not a single mention of `STRef` in the file.

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:163]: Nice!

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs [hunk 2]
```diff
@@ -222,9 +223,9 @@ type CekEvaluationException uni fun = CekEvaluationExceptionCarrying (Term Name
 type CekM cost uni fun s = CekCarryingM cost (Term Name uni fun ()) uni fun s
 
 -- | A budgeting mode to execute an evaluator in.
-data ExBudgetMode cost uni fun = ExBudgetMode
-    { _exBudgetModeSpender :: CekBudgetSpender cost uni fun  -- ^ A spending function.
-    , _exBudgetModeInitSt  :: cost                           -- ^ An initial state.
+data ExBudgetMode cost uni fun s = ExBudgetMode
+    { _exBudgetModeSpender  :: CekBudgetSpender cost uni fun s  -- ^ A spending function.
+    , _exBudgetModeGetFinal :: ST s cost                        -- ^ For accessing the final state.
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:228]: This actually lets you get the state at any time, right? It's just a way to get the state (which we happen to only use at the end).

**Review by michaelpj — APPROVED** [review]
I think this seems sensible and is an improvement, so I say merge it once you're happy.

**Comment by effectfully** [comment]
I played with adding an `UNPACK` pragma and some strictness annotation and got the following results (compared to `StateT` in `CekM`):

```
crowdfunding/1           3.438 ms	3.198 ms    -7.0%
crowdfunding/2           3.490 ms	3.206 ms    -8.1%
crowdfunding/3           3.492 ms	3.212 ms    -8.0%
crowdfunding/4           2.093 ms	1.946 ms    -7.0%
crowdfunding/5           2.095 ms	1.947 ms    -7.1%
future/1                 2.120 ms	1.948 ms    -8.1%
future/2                 4.056 ms	3.773 ms    -7.0%
future/3                 4.050 ms	3.772 ms    -6.9%
future/4                 5.413 ms	4.995 ms    -7.7%
future/5                 6.381 ms	5.935 ms    -7.0%
future/6                 5.450 ms	5.053 ms    -7.3%
future/7                 6.406 ms	5.928 ms    -7.5%
multisigSM/1             4.001 ms	3.715 ms    -7.1%
multisigSM/2             4.018 ms	3.766 ms    -6.3%
multisigSM/3             4.059 ms	3.827 ms    -5.7%
multisigSM/4             4.096 ms	3.864 ms    -5.7%
multisigSM/5             4.765 ms	4.421 ms    -7.2%
multisigSM/6             3.998 ms	3.712 ms    -7.2%
multisigSM/7             4.021 ms	3.756 ms    -6.6%
multisigSM/8             4.064 ms	3.823 ms    -5.9%
multisigSM/9             4.099 ms	3.860 ms    -5.8%
multisigSM/10            4.773 ms	4.420 ms    -7.4%
vesting/1                3.507 ms	3.220 ms    -8.2%
vesting/2                3.097 ms	2.863 ms    -7.6%
vesting/3                3.506 ms	3.249 ms    -7.3%
marlowe/trustfund/1      11.01 ms	10.20 ms    -7.4%
marlowe/trustfund/2      8.750 ms	8.159 ms    -6.8%
marlowe/zerocoupon/1     11.19 ms	10.44 ms    -6.7%
marlowe/zerocoupon/2     8.050 ms	7.574 ms    -5.9%
```

Just using `StrictData` + `STRef`s made things worse though (compared to `StateT`):

```
crowdfunding/1           3.438 ms	3.429 ms    -0.3%
crowdfunding/2           3.490 ms	3.433 ms    -1.6%
crowdfunding/3           3.492 ms	3.431 ms    -1.7%
crowdfunding/4           2.093 ms	2.059 ms    -1.6%
crowdfunding/5           2.095 ms	2.058 ms    -1.8%
future/1                 2.120 ms	2.079 ms    -1.9%
future/2                 4.056 ms	4.041 ms    -0.4%
future/3                 4.050 ms	4.042 ms    -0.2%
future/4                 5.413 ms	5.333 ms    -1.5%
future/5                 6.381 ms	6.291 ms    -1.4%
future/6                 5.450 ms	5.338 ms    -2.1%
future/7                 6.406 ms	6.282 ms    -1.9%
multisigSM/1             4.001 ms	3.945 ms    -1.4%
multisigSM/2             4.018 ms	3.976 ms    -1.0%
multisigSM/3             4.059 ms	4.038 ms    -0.5%
multisigSM/4             4.096 ms	4.082 ms    -0.3%
multisigSM/5             4.765 ms	4.662 ms    -2.2%
multisigSM/6             3.998 ms	3.943 ms    -1.4%
multisigSM/7             4.021 ms	3.973 ms    -1.2%
multisigSM/8             4.064 ms	4.035 ms    -0.7%
multisigSM/9             4.099 ms	4.082 ms    -0.4%
multisigSM/10            4.773 ms	4.663 ms    -2.3%
vesting/1                3.507 ms	3.422 ms    -2.4%
vesting/2                3.097 ms	3.047 ms    -1.6%
vesting/3                3.506 ms	3.423 ms    -2.4%
marlowe/trustfund/1      11.01 ms	10.82 ms    -1.7%
marlowe/trustfund/2      8.750 ms	8.540 ms    -2.4%
marlowe/zerocoupon/1     11.19 ms	11.11 ms    -0.7%
marlowe/zerocoupon/2     8.050 ms	7.896 ms    -1.9%
```

(worse because just using `STRef`s give a bigger speedup than what we see above).

I'm now going to merge the PR, but I've made some minor refactoring changes, so feel free to look.

**Comment by kwxm** [comment]
Here's the validation times on the server (master on left, this branch on right).

```
crowdfunding/1           3.021 ms	2.906 ms    -3.8%
crowdfunding/2           3.058 ms	2.903 ms    -5.1%
crowdfunding/3           3.055 ms	2.899 ms    -5.1%
crowdfunding/4           1.888 ms	1.785 ms    -5.5%
crowdfunding/5           1.884 ms	1.783 ms    -5.4%
future/1                 1.850 ms	1.738 ms    -6.1%
future/2                 3.550 ms	3.373 ms    -5.0%
future/3                 3.544 ms	3.375 ms    -4.8%
future/4                 4.789 ms	4.495 ms    -6.1%
future/5                 5.602 ms	5.333 ms    -4.8%
future/6                 4.782 ms	4.540 ms    -5.1%
future/7                 5.608 ms	5.329 ms    -5.0%
multisigSM/1             3.534 ms	3.344 ms    -5.4%
multisigSM/2             3.538 ms	3.371 ms    -4.7%
multisigSM/3             3.575 ms	3.431 ms    -4.0%
multisigSM/4             3.601 ms	3.460 ms    -3.9%
multisigSM/5             4.180 ms	3.954 ms    -5.4%
multisigSM/6             3.547 ms	3.337 ms    -5.9%
multisigSM/7             3.553 ms	3.383 ms    -4.8%
multisigSM/8             3.603 ms	3.438 ms    -4.6%
multisigSM/9             3.626 ms	3.477 ms    -4.1%
multisigSM/10            4.208 ms	3.979 ms    -5.4%
vesting/1                3.064 ms	2.882 ms    -5.9%
vesting/2                2.715 ms	2.576 ms    -5.1%
vesting/3                3.056 ms	2.883 ms    -5.7%
marlowe/trustfund/1      9.786 ms	9.245 ms    -5.5%
marlowe/trustfund/2      7.759 ms	7.394 ms    -4.7%
marlowe/zerocoupon/1     9.971 ms	9.461 ms    -5.1%
marlowe/zerocoupon/2     7.157 ms	6.827 ms    -4.6%
```

.. and the nofib examples just to make sure.

```
clausify/formula1        144.3 ms	133.8 ms    -7.3%
clausify/formula2        179.5 ms	165.4 ms    -7.9%
clausify/formula3        495.9 ms	457.7 ms    -7.7%
clausify/formula4        678.2 ms	622.9 ms    -8.2%
clausify/formula5        3.207 s	2.982 s     -7.0%
knights/4x4              358.7 ms	316.9 ms    -11.7%
knights/6x6              1048 ms	936.7 ms    -10.6%
knights/8x8              1.789 s	1.583 s     -11.5%
primetest/05digits       224.8 ms	186.7 ms    -16.9%
primetest/08digits       411.9 ms	340.3 ms    -17.4%
primetest/10digits       577.4 ms	476.7 ms    -17.4%
primetest/20digits       1147 ms	952.2 ms    -17.0%
primetest/30digits       1.661 s	1.380 s     -16.9%
primetest/40digits       2.228 s	1.846 s     -17.1%
primetest/50digits       2.190 s	1.812 s     -17.3%
queens4x4/bt             60.45 ms	54.19 ms    -10.4%
queens4x4/bm             78.00 ms	69.94 ms    -10.3%
queens4x4/bjbt1          74.56 ms	67.05 ms    -10.1%
queens4x4/bjbt2          79.82 ms	71.76 ms    -10.1%
queens4x4/fc             188.8 ms	171.5 ms    -9.2%
queens5x5/bt             815.2 ms	726.1 ms    -10.9%
queens5x5/bm             885.9 ms	792.4 ms    -10.6%
queens5x5/bjbt1          953.9 ms	855.3 ms    -10.3%
queens5x5/bjbt2          1011 ms	904.1 ms    -10.6%
queens5x5/fc             2.402 s	2.183 s     -9.1%
```

**Comment by effectfully** [comment]
Thanks! `primetest` is quite impressive.

**Pull request merged**